internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+	if requestNumSamples == 0 {
+		requestNumSamples = defaultRequestNumSamples
+	}
+	requestRetryInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_REQUEST_RETRY_INTERVAL"))
+	if requestRetryInterval == 0 {
+		requestRetryInterval = defaultRequestRetryInterval
+	}
+
+	newRemoteWrite := RemoteWrite{
+		deviceID:             deviceID,
+		tsdb:                 tsdbInstance,
+		waitInterval:         waitInterval,
+		RangeDuration:        RangeDuration,
+		RequestNumSamples:    requestNumSamples,
+		RequestRetryInterval: requestRetryInterval,
+		lastWriteFile:        path.Join(dataDir, LastWriteFileName),
+	}
+
+	lastWriteBytes, err := ioutil.ReadFile(newRemoteWrite.lastWriteFile)
+	if err == nil {
+		if len(lastWriteBytes) != 0 {
+			i, err := strconv.ParseInt(string(lastWriteBytes), 10, 64)
+			if err == nil {
+				newRemoteWrite.LastWrite = time.Unix(i/1000000000, i%1000000000)
+			} else {
+				log.Error(err)
+			}
+		}
+	} else if !os.IsNotExist(err) {
+		log.Error(err)
+	}
+
+	return &newRemoteWrite
+}
+
+// RemoteRecoverableError
+// 'remote.RecoverableError' fields are private and can't be instantiated during testing. Therefore, we use this wrapper
+type RemoteRecoverableError struct {
+	error
+}
+
+func (e RemoteRecoverableError) Error() string {
+	if e.error != nil {
+		return e.Error()
+	} else {
+		return ""
+	}
+}
+
+// WriteClient
+// an interface used for testing
+//go:generate mockgen -package=metrics -destination=write_client_mock.go . WriteClient
+type WriteClient interface {
+	Write(context.Context, []byte) error
+}
+
+type writeClient struct {
+	client remote.WriteClient
+}
+
+func (w *writeClient) Write(ctx context.Context, data []byte) error {
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
+	serverTimeout, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_SERVER_TIMEOUT"))
+	if serverTimeout == 0 {
+		serverTimeout = time.Second * 10
+	}
+
+	// create configuration
+	r.config = RemoteWriteConfiguration{
+		Url:     serverURL,
+		Timeout: serverTimeout,
+	}
+
+	if r.config.Url == "" { // feature is disabled
+		return nil
+	}
+
+	// create client
+	if r.Client == nil {
+		serverURL, err := url.Parse(r.config.Url)
+		if err != nil {
+			return err
+		}
+		client, err := remote.NewWriteClient(r.deviceID, &remote.ClientConfig{
+			Timeout: model.Duration(r.config.Timeout),
+			URL: &config_util.URL{
+				URL: serverURL,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		r.Client = &writeClient{
+			client: client,
+		}
+	}
+
+	return nil
+}
+
+func (r *RemoteWrite) Init(config models.DeviceConfigurationMessage) error {
+	err := r.Update(config)
+	go r.writeRoutine()
```

Comment:
if the feature is disabled (https://github.com/project-flotta/flotta-device-worker/pull/115/files/3a8d1de8e000a338a61f1e57bc29c1924ab0e1ab#diff-51602231cf99bc104842e148b00d32af4ed8812fbf41d942e7b9ed9417ee048dR145), that go routine should not be started.

Reply:
most chances are that it will be enabled. if disabled then probably due to some temporary problem. so I think it is easier to start it and let it operate/not-operate based on config. In general, I'm quite certain that we have other routines that are started even when the feature is disabled. We also create quite a lot of goroutines. I don't think another one is should be a problem. anyhow, I'll handle it in the DeviceConfig story.

Reply:
I highly disagree here, we need to think that it's device level, and maybe those devices are using solar panels, batteries, etc.. having another goroutine can consume more electricity and we should avoid it.

Reply:
The devices won't be anything less than a reasonable computer. look at our code. it is more than a small device can handle.
if this goroutine executes once every 5 minute it's nothing. Besides, if that is the case then entire code should be reviewed. For example. The metric targets. each is a goroutine instead of one. These create context switching and that means that CPU registers are written with new data and that means consuming electricity. The same goes for the node exporter that uses its own process. The same goes for our over frequent GET configuration and heartbeats. Let's be reasonable. We're not talking about a device operated with batteries nor do we have any idea what kind of devices we have.
And I already wrote it would be handled in the Device Config story (https://issues.redhat.com/browse/ECOPROJECT-555). Cause there we'll finalize who creates what first.

Reply:
@eloycoto @jakub-dzon 
I think I confused you.
There's no enable/disable for this feature. It is always enabled. The epic handles setting the URL.
I added this option for development. Similar to OBC_AUTO_CREATE in 'make run' of operator. Only here I have the privilege to do it without changing the makefile.

Reply:
The feature is disabled if the URL is empty and in that case the go routine should not be started. 
Besides, I think it is a very valid use case for users to opt-out from collecting metrics on the cluster side. They might not want to maintain unused Thanos instance for the sake of running edge devices.

Reply:
if it is a valid case then we have to handle disabling the metrics in the device. at the moment we do not. TSDB starts at least one routine. If I am not mistaken, node exporter is started by default. One target is one routine. Also the service and node exporter process.

Reply:
You're right about TSDB. node exporter is down when system metrics are disabled.

Reply:
Right now there is a switch for turning the system metrics off and maybe we could add similar switch for workload metrics. When they would both be off, we could turn TSDB off.

Reply:
ok. so when we'll do ALL of that we can do this one as well. till then no reason for this feature to be required to do more than the rest.